### ISSUE TITLE
How SHAP handle multi-collinearity

### ISSUE THREAD

**Issue by mchirsa5** [shap/shap#1120]
Hi,

I have just started learning XGBoost model. I used SHAP as a tool for feature selection for my XGBoost prediction model. After obtaining the feature importance, I noticed that the SHAP values of some features are equal to zero. After investigating into more details, I found that the features with zero SHAP value are collinear. For example, if features A and B are highly correlated, the SHAP value of B is then set to zero.

However, from my understand, Shapley value uses cooperative game theory to compute the contribution of each signal. So, in case of multicollinearity, shouldn't A and B have the same value of Shapley value?

Also, if the Shapley value can handle multicollinearity, how does SHAP select the signal and ignore the other correlated features?

**Comment by alexcoca** [comment]
Have you tried computing the shap interaction values for your model?

**Comment by LEMTideman** [comment]
Hi @mchirsa5, the question you ask about correlated features is a tricky one, and the phenomenon you observe (collinear features being assigned a SHAP importance score of zero) is actually quite common (and, arguably, problematic) in the field of explainable AI/interpretable ML: it is sometimes referred to by the name of correlation bias.  

SHAP is a post-hoc model-agnostic interpretability method that uses Shapley values from game theory to estimate the predictive importance (i.e. SHAP score) of the features of a machine learning model. When using SHAP, remember that we are explaining the machine learning model (rather than the data!!!). Correlation bias occurs because of how the machine learning algorithm trains the model, not because of how SHAP estimates feature importance. For example, when presented with a group of highly correlated features {A,B,C}, the machine learning algorithm will assign a large weight to one arbitrary representative of the group, let’s say A. Indeed, features B and C hardly provide any more information than A does (they are very similarly associated to the prediction target labels). Features B and C are redundant. The model’s decision-making process will therefore rely heavily on feature A, but not on features B and C. Consequently, features B and C will score poorly when SHAP is used to explain the model. If SHAP were to assign high importance scores to features B and C, it would not be faithful to the model!

**Comment by mchirsa5** [comment]
@alexcoca Sorry for my delay response. I haven't tried plotting the SHAP interaction yet. I will try plotting it. Thank you for the recommendation :)

@LEMTideman Thank you for your response. Please correct me if I misunderstood it. I do apologize for asking such a basic question. I'm very new to machine learning. So, the reason why the redundant variables are set to zero is because they do not exist in the machine leaning model.

**Comment by LEMTideman** [comment]
@mchirsa5: the SHAP values of redundant variables are zero because the model does not depend on them (these variables have no influence on the model's output)

**Comment by mchirsa5** [comment]
@LEMTideman I see. Thank you so much! I have one more question related to the interpretability. I've tried to compare the results between SHAP and correlation analysis. The Spearman's correlation coefficient was calculated between the features (input signals) and the target (predicted value). 

What I have found is that some variables, which were selected by SHAP, have a very low correlation coefficient with the predicted value. Do you know why SHAP selects the signal with very low correlation coefficients? As I understand, SHAP measure the contribution of that signal on the output variable. However, if the signal has a very low correlation coefficient, this signal should have less contribution though?

**Comment by LEMTideman** [comment]
Hi @mchirsa5, you are comparing two different approaches. By calculating the Spearman’s correlation coefficient between the features and target, you are studying the data. By calculating the Shapley values using SHAP, you are studying the model. 

SHAP will tell you which features were considered important by the machine learning algorithm used to train your XGBoost model. SHAP measures the influence that each feature has on the XGBoost model’s prediction, which is not (necessarily) the same thing as measuring correlation. Spearman’s correlation coefficient only takes monotonic relationships between variables into account, whereas SHAP can also account for non-linear non-monotonic relationships between variables. Depending on your XGBoost model, it is possible for a feature whose Spearman’s correlation coefficient is small to have a large impact on the model’s decision-making process. For example, a feature that is not highly correlated with the target may add useful information to the prediction task, assuming the presence of several other features (hence, in a specific subset of the feature space).

**Comment by mchirsa5** [comment]
@LEMTideman Thank for the clarification!!

**Comment by dushyant-007** [comment]
Hi, @LEMTideman  Actually my problem is what you said in your last comment. My correlation analysis shows less importance of a feature and a P-value of 0.65. Ideally it means that I can also remove this feature from my model. However, Its impact on my model is pretty high (from SHAP). So, for validation I made another model without that particular feature, and my accuracy actually increased (Sample size is limited , so removing the variable had a very positive impact). I cannot understand why SHAP is giving this output. 

One possible reason could be that this particular feature's value changes from 0.01 to 45 (4500x). while other feature do not change more than 3x throughout the range. But I am not able to connect the dots as to why that would cause this problem to happen? 

However the multi-collinearity problem doesn't exist as the input variables are very scattered and doesn't really have any relation.

Thanks for any response.

**Comment by GZuin** [comment]
@dushyant-007 I'm a little late to the party, but anyway. As @LEMTideman pointed, SHAP values are a measure of the importance of a feature in relation to the model. This is different from the true importance of a feature.

Consider a dataset that contains the index columns, and also consider that the model overfits and learns to use the index value to make a prediction. You can expect the shap values of the index to be through the roof as that feature alone is responsible for the prediction. However, out-of-sample accuracy will be rock bottom because that is not a feature that is useful for actual prediction. If you remove the index feature from the train data, you will train a new model that will have a different shap profile. And in this case, test accuracy should improve.

Shap is not a measure of "how important a given feature is in the real world", it is simply "how important a feature is to the model". If your model is not trained properly (for instance, when data cleanup is not employed), then the shap values will be distant from what you would expect. On the other hand, if you guarantee that your model is correct, then shap values should be a good approximation of what reality is like.

**Comment by dushyant-007** [comment]
@GZuin thanks for the comment. Actually, there is a problem a problem with my code. So here's what happened - 

I created the model. It performs great. But I did standardization of training set. Did the same for test set, so accuracy was high. But while passing the variable to SHAP I passed the original values. Obviously, since the friction factor had values from 0.01 to 45. It somehow still mattered , the rest didn't (for the model) and hence the result.

**Comment by scoste** [comment]
@LEMTideman am I correct in thinking that although SHAP should test all combinations of feature coalitions, only in a few of these will the marginal contribution of B be evaluated between coalitions lacking A or C. In those cases, B will shine, but in most cases B's marginal contribution will be calculated using coalitions that contain A and/or C, and in those cases the algorithm will get what signal B has in it from A and/or C instead, thus greatly reducing the marginal contribution of B. And is that why, even though you are in theory looking at models trained on all possible feature combinations, a slightly better feature will tend to eclipse any collinear features?

To me, this would also suggest that the weighted average of B's importance could never go truly to 0, since those cases where it's contribution is calculated without A or C would still yield an above-zero value for useful features. So I assume the zero values mentioned above are only approximately zero—otherwise my understanding needs work.

Hope this makes sense. Thanks!

**Comment by jessegmeyerlab** [comment]
I found this discussion very useful. Thanks all!

**Comment by ashishthanki** [comment]
> feature importance, I noticed that the SHAP values of some features are equal to zero. After investigating into more details, I found that the features with zero SHAP value are collinear. For example, if features A and B are highly correlated, the SHAP value of B is then set to zero.
> 
> However, from my understand, Shapley value uses cooperative game theory to compute the con

If you are using a XGBoost model with features which are possibly correlated another approach would be to introduce subsampling within XGBoost itself. Check [this comment](https://github.com/slundberg/shap/issues/288#issuecomment-460476216). 

The model has a family of parameters for subsampling of columns - see [this](https://xgboost.readthedocs.io/en/stable/parameter.html#parameters-for-tree-booster). You can tune each of them which define the ratio of the number of features you would like to use during training.

Specifically, the following: `colsample_bytree, colsample_bylevel, colsample_bynode` should help handle the multicollinearity amongst your features.

**Comment by SaraAbbaspour** [comment]
Hi,

I am also using XGBoost as my classifier. I have numerical (i.e., age) and categorical variables (e.g., race/ethnicity, sex, etc) as input features. I am using SHAP to find the effect of each of the input variables (their amplitude and direction) on the model’s output. Some of the variables in my dataset are highly correlated. for example, I have 4 categories for race/ethnicity (4 binary 0/1 columns) in the dataset and I should remove one of them because of collinearity.
Now, I am wondering if I remove the correlated variable, how I will be able to see the effect (the direction) of that variable on the output.
I tested the result with and without the colinear variables, the performance of the classifier is almost the same and a the SHAP plot changes slightly.

**Comment by ashishthanki** [comment]
> 



> Hi,
> 
> I am also using XGBoost as my classifier. I have numerical (i.e., age) and categorical variables (e.g., race/ethnicity, sex, etc) as input features. I am using SHAP to find the effect of each of the input variables (their amplitude and direction) on the model’s output. Some of the variables in my dataset are highly correlated. for example, I have 4 categories for race/ethnicity (4 binary 0/1 columns) in the dataset and I should remove one of them because of collinearity. Now, I am wondering if I remove the correlated variable, how I will be able to see the effect (the direction) of that variable on the output. I tested the result with and without the colinear variables, the performance of the classifier is almost the same and a the SHAP plot changes slightly.

I would suggest using colsample_bytree, colsample_bylevel, colsample_bynode as they will help handle the collinearity amongst your features. This prevents all your trees from using all the features and trains the model on a subset of features within the dataset.

**Comment by SaraAbbaspour** [comment]
> 

Thank you for your suggestion. 
I have 16 features, if 3 of them are collinear with other features, does it mean that the value for colsample_bytree, colsample_bylevel, and colsample_bynode should be 0.8?

**Comment by JunqiZhao** [comment]
> Hi @mchirsa5, the question you ask about correlated features is a tricky one, and the phenomenon you observe (collinear features being assigned a SHAP importance score of zero) is actually quite common (and, arguably, problematic) in the field of explainable AI/interpretable ML: it is sometimes referred to by the name of correlation bias.
> 
> SHAP is a post-hoc model-agnostic interpretability method that uses Shapley values from game theory to estimate the predictive importance (i.e. SHAP score) of the features of a machine learning model. When using SHAP, remember that we are explaining the machine learning model (rather than the data!!!). Correlation bias occurs because of how the machine learning algorithm trains the model, not because of how SHAP estimates feature importance. For example, when presented with a group of highly correlated features {A,B,C}, the machine learning algorithm will assign a large weight to one arbitrary representative of the group, let’s say A. Indeed, features B and C hardly provide any more information than A does (they are very similarly associated to the prediction target labels). Features B and C are redundant. The model’s decision-making process will therefore rely heavily on feature A, but not on features B and C. Consequently, features B and C will score poorly when SHAP is used to explain the model. If SHAP were to assign high importance scores to features B and C, it would not be faithful to the model!

Hi @LEMTideman, thanks for the informative discussion! I wonder would the zero SHAP score happen when we adopt the "observational" approach (which is "true to the data") for calculating the SHAP score? It seems the correlation bias happens when we use the interventional approach (which is "true to the model" if I understand correctly).

**Comment by dankirk95** [comment]
I would appreciate clarity on one of the discussion points here. It is stated in @LEMTideman 's excellent response that "when presented with a group of highly correlated features {A,B,C}, the machine learning algorithm will assign a large weight to one arbitrary representative of the group".

However, I believe this is not always the case. For example, it is described [here](https://explained.ai/rf-importance/) how RF can either assign a high feature importance to one of a set of correlated features and 0 to the rest (as described by @LEMTideman ) or spread the importance of out across the correlated variables, i.e., if they are all decent predictors of the outcomes, feature importance for all of them is underreported instead of assigning them all a similarly high feature importance. 

This difference occurs due to the choice of mean decrease in impurity (in the former) or information gain (in the latter) as the [performance criterion](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html).

Hence, if "SHAP will tell you which features were considered important by the machine learning algorithm" (as stated above) then presumably that means SHAP will also be affected by the choice of criterion here in the model configuration. 

This will have consequences on the interpretation of which features are important since if feature importance is averaged across a group of correlated features, then one might get the impression that these features are of little importance to the model. 

Am I mistaken in my understanding here?

**Comment by dankirk95** [comment]
Really helpful discussion here and I appreciate the answers given above. One thing that I think could be clearer is 

> "when presented with a group of highly correlated features {A,B,C}, the machine learning algorithm will assign a large weight to one arbitrary representative of the group, let’s say A ... The model’s decision-making process will therefore rely heavily on feature A, but not on features B and C"

This wording suggests that all ML algorithms will do this, although I think this is dependent on the algorithm used and even the scoring criteria within different version of the same algorithm ([shown here](https://explained.ai/rf-importance/)). This would suggest that the selected scoring criteria for certain algorithms (e.g., gini vs entropy vs log_loss for RF) would strongly influence the results of SHAP; yet, this is not discussed above. 

Am I correct in thinking this?

**Comment by jesuinovieira** [comment]
@CloseChoice, I believe this thread can be labeled as "good reads" (#3887)
